### ISSUE TITLE
Dev instance: Fedora Login fails since "module 'base64' has no attribute 'encodestring'"

### ISSUE THREAD

**Issue by oturpe** [fedora-infra/anitya#1145]
I set up a local development instance using the Vagrant method.
The web page opens fine at http://127.0.0.1:5000/
but Fedora login produces the following error.

This is caused by `social-storage-sqlalchemy` not supporting Python 3.9 or later.
I logged  python-social-auth/social-storage-sqlalchemy#6.
In the mean time, I suppose I have to hack my dev instance
so that it allows the things I am trying to do without login.

> AttributeError: module 'base64' has no attribute 'encodestring'
>
> Traceback (most recent call last)
>    File "/usr/lib/python3.9/site-packages/flask/app.py", line 2464, in __call__
>    return self.wsgi_app(environ, start_response)
>    File "/usr/lib/python3.9/site-packages/flask/app.py", line 2450, in wsgi_app
>    response = self.handle_exception(e)
>    File "/usr/lib/python3.9/site-packages/flask/app.py", line 1867, in handle_exception
>    reraise(exc_type, exc_value, tb)
>    File "/usr/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
>    raise value
>    File "/usr/lib/python3.9/site-packages/flask/app.py", line 2447, in wsgi_app
>    response = self.full_dispatch_request()
>    File "/usr/lib/python3.9/site-packages/flask/app.py", line 1952, in full_dispatch_request
>    rv = self.handle_user_exception(e)
>    File "/usr/lib/python3.9/site-packages/flask/app.py", line 1821, in handle_user_exception
>    reraise(exc_type, exc_value, tb)
>    File "/usr/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
>    raise value
>    File "/usr/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request
>    rv = self.dispatch_request()
>    File "/usr/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request
>    return self.view_functions[rule.endpoint](**req.view_args)
>    File "/usr/local/lib/python3.9/site-packages/social_flask/utils.py", line 43, in wrapper
>    return func(backend, *args, **kwargs)
>    File "/usr/local/lib/python3.9/site-packages/social_flask/routes.py", line 14, in auth
>    return do_auth(g.backend)
>    File "/usr/local/lib/python3.9/site-packages/social_core/actions.py", line 29, in do_auth
>    return backend.start()
>    File "/usr/local/lib/python3.9/site-packages/social_core/backends/base.py", line 34, in start
>    if self.uses_redirect():
>    File "/usr/local/lib/python3.9/site-packages/social_core/backends/open_id.py", line 234, in uses_redirect
>    return self.openid_request().shouldSendRedirect()
>    File "/usr/local/lib/python3.9/site-packages/social_core/backends/open_id.py", line 239, in openid_request
>    return self.consumer().begin(url_add_parameters(self.openid_url(),
>    File "/usr/local/lib/python3.9/site-packages/openid/consumer/consumer.py", line 360, in begin
>    return self.beginWithoutDiscovery(service, anonymous)
>    File "/usr/local/lib/python3.9/site-packages/openid/consumer/consumer.py", line 383, in beginWithoutDiscovery
>    auth_req = self.consumer.begin(service)
>    File "/usr/local/lib/python3.9/site-packages/openid/consumer/consumer.py", line 611, in begin
>    assoc = self._getAssociation(service_endpoint)
>    File "/usr/local/lib/python3.9/site-packages/openid/consumer/consumer.py", line 1184, in _getAssociation
>    self.store.storeAssociation(endpoint.server_url, assoc)
>    File "/usr/local/lib/python3.9/site-packages/social_core/store.py", line 25, in storeAssociation
>    self.assoc.store(server_url, association)
>    File "/usr/local/lib/python3.9/site-packages/social_sqlalchemy/storage.py", line 218, in store
>    assoc.secret = base64.encodestring(association.secret).decode()
>    AttributeError: module 'base64' has no attribute 'encodestring'

**Comment by LenkaSeg** [comment]
That is a known issue. Encodestring was deprecated and is now encodebytes.
There is a [pull request](https://github.com/python-social-auth/social-storage-sqlalchemy/pull/5) hanging on social-sqlalchemy for 6 months with no response.
To avoid this problem [this ticket](https://github.com/fedora-infra/anitya/issues/1139) was created.

**Comment by oturpe** [comment]
> That is a known issue. Encodestring was deprecated and is now encodebytes.
> There is a [pull request](https://github.com/python-social-auth/social-storage-sqlalchemy/pull/5) hanging on social-sqlalchemy for 6 months with no response.
> To avoid this problem [this ticket](https://github.com/fedora-infra/anitya/issues/1139) was created.

I guess it is safe to say "upstream is dead"
if a one line pull request does not get resolved in that much time.
For anybody suffering for this problem,
here is how I anyhow managed to run the development instance:

1. Removed any auth related annotations from backend functions I needed to call:
  ```
  #@authentication.require_token
  def post(self):
  ```
2. Replaced places where login user name was used with dummy names:
  ```
  project = utilities.create_project(
                Session,
                #user_id=flask_login.current_user.email,
                user_id="TESTDEV"
```
3. In ui templates, removed any conditionals that were hiding components only inteded for logged in users:
  ```
  <!-- {%- if user and user.is_authenticated -%}-->
  {%- if user }
  <li class="active">
  ```

**Comment by Zlopez** [comment]
Thanks for sharing this, it will help to get over authentication in dev env. But we need to find more permanent solution for production instance. I would be glad if we could preserve same options for login as we have now, but I'm not sure if anybody even used something other than Fedora account.

**Comment by Zlopez** [comment]
> That is a known issue. Encodestring was deprecated and is now encodebytes.
> There is a [pull request](https://github.com/python-social-auth/social-storage-sqlalchemy/pull/5) hanging on social-sqlalchemy for 6 months with no response.
> To avoid this problem [this ticket](https://github.com/fedora-infra/anitya/issues/1139) was created.

The pull request was merged today, let's hope there will be a new version out soon.

**Comment by Zlopez** [comment]
I deployed a hotfix on https://stg.release-monitoring.org and it seems to work fine.

I will prepare Anitya for new release and deploy the hotfix to release-monitoring.org as well. Hopefully, this will happen soon.

I also introduced the hotfix to dev environments we have, so they should work as intended.

Closing the issue, because this is no longer the case and we will migrate to another authentication system in future.
